Folder locations mk2

JacksonRG opened the pull request:
closes #4590
# Enhancement
Currently openshot's memory of folder locations is tied to projects. Some users may keep lots (or all) of their resource files in the same folder. If they do, this change will start them off in the folder they last imported from. If not, this can be switched off in `Settings` > `Locations`

# Example
1) Open a project
2) Start to import a file. (Ctrl+f)
3) Go into a sub folder and choose an image to import.
3) Create a new project
4) start to Import another file (Ctrl+f)

Notice that you are not in the same subfolder.

# Test
Repeat the steps from example, and on the last step, You should be in the subfolder navigated to in step 3. (Unless you've changed `Settings` > `Locations`)

Commit:
Add settings for import, export, and projects

Created setting for where file explorers open.
Either in the directory of the current project,
or in the directory where the last file was either imported, exported, or saved/loaded

Commit:
Link folder operations to new settings

- Connect all import, export, and open/save actions to the settings.
- Corrected logic error in settings.

JacksonRG commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Import/Export Folders in Settings](https://trello.com/c/ABZbyHRd/48-import-export-folders-in-settings)

Colorjet3 commented:
What was the reason PR #4757 closed?  I thought having at minimum those three file paths in the Locations tab is great.  Here is my thoughts on this:
1.  PR #4757 should be implemented.
2.  The user has the option to populate any of the fields.
3.  If "File Imports:" = /home/jackson/Videos/Resources, then regardless of where the last import was made, OpenShot always defaults to this path for imports.  However, if this field is left empty, then OpenShot resorts to the above logic and defaults to the last location that an import was made.
4. This logic should apply to "Video Exports" and "Save Projects" (as I suggested this should be "Open/Save Projects" for clarity).

This gives the user the option/freedom to influence the behavior of "File Imports", "Video Exports", and "Open/Save Projects".

JacksonRG commented:
@Colorjet3 In talking it over with @jonoomph we realized that if the user does have a specific folders they always import from and export to, this logic will work for them without them having to find and change the setting. The only difference would be navigating back if they import from a different folder.

I wouldn't mind adding the folder settings back in. But for now, changing the default behavior benefits all users, not just the ones who find out about setting the folders.

jonoomph commented on src/classes/settings.py:
```diff
@@ -39,6 +41,18 @@ class SettingStore(JsonDataStore):
 
+    @unique
+    class pathType(Enum):
+        RECENT=0 # Return the last used path for the same action
```
Is there a reason we are starting with 1 here, but using auto() in other places?

JacksonRG replied:
My reason was just to ensure that the value of RECENT and PROJECT were the same as the value of their respective settings.

jonoomph replied:
But we could change the settings to use 1, and 2 as values, right?

JacksonRG replied:
Yeah, No reason the setting values can't be 1 and 2

jonoomph commented on src/classes/settings.py:
```diff
@@ -39,6 +41,18 @@ class SettingStore(JsonDataStore):
+        PROJECT=1 # Return the current project path
+
+    @unique
+    class actions(Enum):
```
action => actionType: what do you think about renaming this to clearly be a 'type'?

jonoomph commented on src/classes/settings.py:
```diff
@@ -111,3 +125,75 @@ def save(self):
+    def setDefaultPath(self, action: pathType, recent_path: str):
+        """ Change the path setting corresponding to the given action """
+        if not os.path.exists(recent_path):
+            return
```
Should we **log** a warning here if the path is invalid?

jonoomph commented on src/classes/settings.py:
```diff
@@ -111,3 +125,75 @@ def save(self):
+            log.debug(f"Setting %s to %s" % (action.value, recent_path))
+            self.set(setting["path"], recent_path)
+        except:
+            log.error(f"Error setting the path of %s" % setting["path"])
```
Will `setting` even be defined at this point in the `except` clause?  Perhaps the log.error could be a safer message, that will definitely work?

JacksonRG replied:
Makes sense. I'll make sure the error msg is safe

jonoomph commented on src/classes/settings.py:
```diff
@@ -111,3 +125,75 @@ def save(self):
+            # Make sure default_path is a directory
+            default_path = os.path.dirname(default_path)
+
+        if (not default_path) or (not os.path.exists(default_path)):
```
I don't think you need the () and ()'s

JacksonRG replied:
@jonoomph what if I simplified it like this: `if not (default_path and os.path.exists(default_path))`  ?

I still think it's good to be clear what the`not` is affecting

jonoomph commented on src/classes/settings.py:
```diff
@@ -111,3 +125,75 @@ def save(self):
+        if action == self.actions.SAVE:
+            log.info("Adding file name for project path")
+            # Default name unless a project is currently active.
+            project_name = os.path.basename(self.app.project.current_filepath or "") or "%s.osp" % _("Untitled Project")
```
This looks sus:
`project_name = os.path.basename(self.app.project.current_filepath or "") or "%s.osp" % _("Untitled Project")`

What does `basename("")` return? Are you trying to prevent `basename(None)`?

jonoomph replied:
Also, I'm not sure I understand adding the `project_name` to this function. It seems like the caller would add on the project_name or default name to the path returned.

jonoomph commented on src/windows/main_window.py:
```diff
@@ -565,22 +564,26 @@ def actionOpen_trigger(self):
         file_path = app.project.current_filepath
         if not file_path:
-            recommended_path = os.path.join(info.HOME_PATH, "%s.osp" % _("Untitled Project"))
+            recommended_path = s.getDefaultPath(settings.actions.SAVE)
```
As mentioned above, I would add on the project_name or "Untitled" name here, instead of inside the function.

jonoomph commented on src/windows/main_window.py:
```diff
@@ -565,22 +564,26 @@ def actionOpen_trigger(self):
                 _("Save Project..."),
                 recommended_path,
                 _("OpenShot Project (*.osp)"))[0]
+            s.setDefaultPath(s.actions.SAVE, file_path)
```
Same: add project name here

jonoomph commented:
This mostly looks good. Just a little bit of clean-up, and some nitpicks. The biggest change is to never return the project name in the `getDefaultPath()` function... let the caller do that. That function is all about the `directory`, and nothing else.

Commit:
Incorporating PR Feedback

- Improved Names
- added error statements
- don't handle naming of files in getDefaultPaths
- don't set defaults if file dialogue is canceled

Commit:
Correct variable name that was missed

jonoomph commented on src/settings/_default.settings:
```diff
@@ -967,5 +967,83 @@
+    "type": "dropdown",
+    "values": [
+      {
+        "value": 0,
```
For example, this could be 1 instead of 0 (so we can use `auto()` in the enum)

jonoomph commented on src/windows/main_window.py:
```diff
@@ -649,18 +658,21 @@ def auto_save_project(self):
-        if not recommended_path:
-            recommended_path = os.path.join(
-                info.HOME_PATH, "%s.osp" % _("Untitled Project"))
+        recommended_folder = s.getDefaultPath(s.actionType.SAVE)
```
This behavior is different now, right? We were using the current_filepath previously, and only using "Untitled Project" when there is no current_filepath. I'm not sure about this though... just curious.

JacksonRG replied:
Fixed this. Now it'll check for a project name

jonoomph commented on src/windows/main_window.py:
```diff
@@ -691,6 +702,8 @@ def actionImportFiles_trigger(self):
             start_location,
             )[0]
 
+        if len(qurl_list):
```
what is this code for? `qurl_list`? Is there a reason we aren't just using our recommended_path from above?

JacksonRG replied:
I figured the most logical `new default` path is the folder of the *last* file selected.

pseudocode:
```
if any files were selected
    setDefault <- <basename of LAST file selected>
```
I'll add a comment to explain that.

Commit:
More feedback improvements

- Use auto for both enums
- If project has a name, use it by default for Save As.
- Comment to explain path set after multiple imports.

jonoomph commented:
LGTM!
